<a href="https://colab.research.google.com/github/carolvieirav/Case_CloudWalk/blob/main/DesafioCloudWalk.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np
import re
import collections
import matplotlib.pyplot as plt
from datetime import datetime

In [2]:
data = pd.read_csv('https://gist.githubusercontent.com/cloudwalk-tests/76993838e65d7e0f988f40f1b1909c97/raw/9ceae962009236d3570f46e59ce9aa334e4e290f/transactional-sample.csv')

In [3]:
data.head()

transaction_id  merchant_id  user_id  ... transaction_amount device_id  has_cbk
0        21320398        29744    97051  ...             374.56  285475.0    False
1        21320399        92895     2708  ...             734.87  497105.0     True
2        21320400        47759    14777  ...             760.36       NaN    False
3        21320401        68657    69758  ...            2556.13       NaN     True
4        21320402        54075    64367  ...              55.36  860232.0    False

[5 rows x 8 columns]

In [4]:
data.info() # Column device_id is the only one with missing values

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3199 entries, 0 to 3198
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   transaction_id      3199 non-null   int64  
 1   merchant_id         3199 non-null   int64  
 2   user_id             3199 non-null   int64  
 3   card_number         3199 non-null   object 
 4   transaction_date    3199 non-null   object 
 5   transaction_amount  3199 non-null   float64
 6   device_id           2369 non-null   float64
 7   has_cbk             3199 non-null   bool   
dtypes: bool(1), float64(2), int64(3), object(2)
memory usage: 178.2+ KB


#Data Pre-processing

In [11]:
def split_date(x):
  '''
  splits date between date and time information
  '''
  y = re.split('T', x)
  return(y)

In [12]:
data['transaction_date'] = data.transaction_date.apply(split_date)

In [17]:
data['date'] = [x[0] for x in data.transaction_date]
data['time'] = [x[1] for x in data.transaction_date]

In [21]:
np.unique(data.duplicated('date'),return_counts=True)

(array([False,  True]), array([  31, 3168]))

In [22]:
data.drop(['transaction_date'], axis=1, inplace=True)
data.head()

transaction_id  merchant_id  user_id  ... has_cbk        date             time
0        21320398        29744    97051  ...   False  2019-12-01  23:16:32.812632
1        21320399        92895     2708  ...    True  2019-12-01  22:45:37.873639
2        21320400        47759    14777  ...   False  2019-12-01  22:22:43.021495
3        21320401        68657    69758  ...    True  2019-12-01  21:59:19.797129
4        21320402        54075    64367  ...   False  2019-12-01  21:30:53.347051

[5 rows x 9 columns]

#Exploratory Analysis

In [23]:
def duplicated_count(x):
  '''
  count duplicated rows
  '''
  y=np.unique(x.duplicated(), return_counts=True)
  print(y[1])

data.apply(duplicated_count)

[3199]
[1756 1443]
[2704  495]
[2925  274]
[3124   75]
[1997 1202]
[   2 3197]
[  31 3168]
[3199]


transaction_id        None
merchant_id           None
user_id               None
card_number           None
transaction_amount    None
device_id             None
has_cbk               None
date                  None
time                  None
dtype: object

Takeaways:

**transaction_id:** every transaction id is unique;

**time:** every transaction time is unique;

**merchant_id:** 1756 merchants have more than one transaction;

**user_id:** 2704 users have more than one transaction;

**card_number:** 2925 card_numers have more than one transaction;

**transaction_amount:** 3124 products have more than one transaction with the same value;

**device_id:** 1202 divices have more than one transaction;

**date:** 3168 dates have more than one transaction.

In [153]:
user_info = data[['user_id','card_number','device_id']]
transaction_info = data[['user_id', 'card_number', 'merchant_id']]

In [173]:
data.apply(duplicated_count)

[2704  495]
[2925  274]
[1997 1202]


user_id        None
card_number    None
device_id      None
dtype: object